core/src/main/java/com/linecorp/armeria/common/stream/TimeoutSubscriber.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.concurrent.ScheduledFuture;
+
+public class TimeoutSubscriber<T> implements Subscriber<T> {
+
+    private final Subscriber<? super T> delegate;
+    private final EventExecutor executor;
+
+    private final StreamTimeoutMode timeoutMode;
+    private ScheduledFuture<?> timeoutFuture;
+
+    private Subscription subscription;
+    private final Duration timeoutDuration;
+
+    public TimeoutSubscriber(Subscriber<? super T> delegate, EventExecutor executor, Duration timeoutDuration, StreamTimeoutMode timeoutMode) {
+        this.delegate = requireNonNull(delegate, "delegate");
+        this.executor = requireNonNull(executor, "executor");
+        this.timeoutDuration = requireNonNull(timeoutDuration, "timeoutDuration");
+        this.timeoutMode = requireNonNull(timeoutMode, "timeoutMode");
+    }
+
+    private ScheduledFuture<?> scheduleTimeout() {
+        return executor.schedule(() -> {
+            subscription.cancel();
+            delegate.onError(new TimeoutException(
+                    String.format("Stream timed out after %d ms (timeout mode: %s)", timeoutDuration.toMillis(), timeoutMode)));
+        }, timeoutDuration.toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        delegate.onSubscribe(s);
+        subscription = s;
+        timeoutFuture = scheduleTimeout();
+    }
+
+    @Override
+    public void onNext(T t) {
+        delegate.onNext(t);
+        if (timeoutFuture.isCancelled()) {
+            return;
+        }
+        switch (timeoutMode) {
+            case UNTIL_NEXT:
+                timeoutFuture.cancel(false);
+                timeoutFuture = scheduleTimeout();
```

Comment:
Note that we need to cancel the timeout future when the stream is aborted though, because we don't want the scheduled tasks linger in the heap even after the stream is gone.